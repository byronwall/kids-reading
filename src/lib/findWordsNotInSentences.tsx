import { type DetailedLearningPlan } from "~/types/models";

export function findWordsNotInSentences(plan: DetailedLearningPlan) {
  // Create a set of all word IDs in sentences for quick lookup
  const sentenceWordIds = new Set<string>();
  plan.sentences.forEach((sentence) => {
    sentence.words.forEach((word) => {
      sentenceWordIds.add(word.id);
    });
  });

  // Filter out words that are not present in any sentence
  const wordsNotInSentences: DetailedLearningPlan["lessons"][0]["words"] = [];

  plan.lessons.forEach((lesson) => {
    lesson.words.forEach((word) => {
      if (!sentenceWordIds.has(word.id)) {
        wordsNotInSentences.push(word);
      }
    });
  });

  return wordsNotInSentences;
}
