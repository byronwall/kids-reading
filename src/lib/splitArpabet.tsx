import { dictionary } from "cmu-pronouncing-dictionary";

export type SyllableInfo = {
  syllables: string[];
  stressLevels: number[];
  arpabet: string;
};

export function getArpabet(word: string): string | undefined {
  const arpabetCode = dictionary[word];

  return arpabetCode;
}

export function getSyllables(word: string): SyllableInfo {
  const arpabetCode = getArpabet(word);

  if (arpabetCode === undefined) {
    throw new Error(`Word not found: ${word}`);
  }

  return splitArpabet(arpabetCode);
}

// rules that appears to be true for splitting into syllables
// all syllable parts must be returned -- don't drop any
// only 1 number = return everything as a single syllable "D AO1 G"

// maybe a 1 tells you to split the syllable before it
// "JH ER1 N IY0" = "JH ER1 | N IY0"
// "OW1 SH AH0 N" = "OW1 | SH AH0 N"
// "P AH0 T EY1 T OW2" = "P AH0 | T EY1 | T OW2"

// vowels together must be split there
// "S K W ER1 AH0 L" = "S K W ER1 | AH0 L"

// lots of consonants together must be split ???
// "D IH0 K T EY1 T ER0 SH IH2 P" = "D IH0 K | T EY1 | T ER0 | SH IH2 P" - between doubles?
// "K AH1 M F ER0 T AH0 B AH0 L" = "K AH1 M | F ER0 | T AH0 | B AH0 L"
// "K AA1 N S T AH0 B AH0 L" = "K AA1 N | S T AH0 | B AH0 L"  - three = split in middle??
// "AH0 T EH1 N SH AH0 N" = "AH0 | T EH1 N | SH AH0 N" - between doubles

export function splitArpabet(arpabet: string): SyllableInfo {
  type SyllablePiece = {
    part: string;
    stressLevel?: number;
  };

  // split on space
  const pieces = arpabet.split(" ").map((c) => c.trim());

  // convert pieces to syllable pieces
  const syllablePieces: SyllablePiece[] = pieces.map((part) => {
    const _stressLevel = Number(part.match(/[0-9]/)?.[0]);
    const stressLevel = isNaN(_stressLevel) ? undefined : _stressLevel;

    // remove last character if stress level
    if (stressLevel !== undefined) {
      part = part.slice(0, -1);
    }

    return { part, stressLevel };
  });

  // put the pieces into groups

  const groups: SyllablePiece[][] = [[]];

  let curGroupHasVowel = false;

  for (let i = 0; i < syllablePieces.length; i++) {
    const prevPiece = syllablePieces[i - 1];
    const piece = syllablePieces[i];
    const nextPiece = syllablePieces[i + 1];

    // determine if current piece should start a new group
    const isVowel = piece?.stressLevel !== undefined;

    if (curGroupHasVowel) {
      const isNextVowel = nextPiece?.stressLevel !== undefined;

      const wasPrevVowel = prevPiece?.stressLevel !== undefined;

      const nextTwoVowels = isVowel || isNextVowel;

      const tripleConsonant = !wasPrevVowel && !isVowel && !isNextVowel;

      if (nextTwoVowels || tripleConsonant) {
        groups.push([]);
        curGroupHasVowel = false;
      }
    }

    const curGroup = groups[groups.length - 1];

    if (curGroup === undefined || piece === undefined) {
      throw new Error(`curGroup is undefined`);
    }

    // if piece has stress level, add it to the group

    curGroup.push(piece);

    if (isVowel) {
      curGroupHasVowel = true;
    }
  }

  const syllables = groups.map((group) => group.map((p) => p.part).join(" "));
  const stressLevels = groups
    .map((group) =>
      group.reduce<number | undefined>((acc, p) => {
        if (acc !== undefined) {
          return acc;
        }

        return p.stressLevel;
      }, undefined)
    )
    .filter((n) => n !== undefined) as number[];

  // create groups of syllables from syllable pieces

  return { syllables, stressLevels, arpabet };
}
