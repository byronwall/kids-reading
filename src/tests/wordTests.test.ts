import { expect, test } from "vitest";

import { getSyllables } from "~/lib/splitArpabet";

// these are words where the syllable break seems off by one
// trying to fix them currently breaks other things in obvious ways.
const trickWords = [
  "believer",
  "unbelievable",
  "stretching",
  "belly",
  "apostle",
  "blackout",
  "timeout",
  "watching",
];

const wordsToTest2 = [
  "believe",
  "comfortable",
  "kitten",
  "couch",
  "dog",
  "potato",
  "loyal",
  "church",
  "throat",
  "squirrel",
  "journey",
  "constable",
  "attention",
  "dictatorship",
  "ocean",
  "happiness",
  "suffering",
];

wordsToTest2.forEach((word) => {
  test("syllables for " + word, () => {
    const syllables = getSyllables(word);
    expect(syllables).toMatchSnapshot();
  });
});
