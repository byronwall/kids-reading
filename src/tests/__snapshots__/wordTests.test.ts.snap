// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syllables for attention 1`] = `
{
  "arpabet": "AH0 T EH1 N SH AH0 N",
  "stressLevels": [
    0,
    1,
    0,
  ],
  "syllables": [
    "AH",
    "T EH N",
    "SH AH N",
  ],
}
`;

exports[`syllables for believe 1`] = `
{
  "arpabet": "B IH0 L IY1 V",
  "stressLevels": [
    0,
    1,
  ],
  "syllables": [
    "B IH",
    "L IY V",
  ],
}
`;

exports[`syllables for church 1`] = `
{
  "arpabet": "CH ER1 CH",
  "stressLevels": [
    1,
  ],
  "syllables": [
    "CH ER CH",
  ],
}
`;

exports[`syllables for comfortable 1`] = `
{
  "arpabet": "K AH1 M F ER0 T AH0 B AH0 L",
  "stressLevels": [
    1,
    0,
    0,
    0,
  ],
  "syllables": [
    "K AH M",
    "F ER",
    "T AH",
    "B AH L",
  ],
}
`;

exports[`syllables for constable 1`] = `
{
  "arpabet": "K AA1 N S T AH0 B AH0 L",
  "stressLevels": [
    1,
    0,
    0,
  ],
  "syllables": [
    "K AA N",
    "S T AH",
    "B AH L",
  ],
}
`;

exports[`syllables for couch 1`] = `
{
  "arpabet": "K AW1 CH",
  "stressLevels": [
    1,
  ],
  "syllables": [
    "K AW CH",
  ],
}
`;

exports[`syllables for dictatorship 1`] = `
{
  "arpabet": "D IH0 K T EY1 T ER0 SH IH2 P",
  "stressLevels": [
    0,
    1,
    0,
    2,
  ],
  "syllables": [
    "D IH K",
    "T EY",
    "T ER",
    "SH IH P",
  ],
}
`;

exports[`syllables for dog 1`] = `
{
  "arpabet": "D AO1 G",
  "stressLevels": [
    1,
  ],
  "syllables": [
    "D AO G",
  ],
}
`;

exports[`syllables for happiness 1`] = `
{
  "arpabet": "HH AE1 P IY0 N AH0 S",
  "stressLevels": [
    1,
    0,
    0,
  ],
  "syllables": [
    "HH AE",
    "P IY",
    "N AH S",
  ],
}
`;

exports[`syllables for journey 1`] = `
{
  "arpabet": "JH ER1 N IY0",
  "stressLevels": [
    1,
    0,
  ],
  "syllables": [
    "JH ER",
    "N IY",
  ],
}
`;

exports[`syllables for kitten 1`] = `
{
  "arpabet": "K IH1 T AH0 N",
  "stressLevels": [
    1,
    0,
  ],
  "syllables": [
    "K IH",
    "T AH N",
  ],
}
`;

exports[`syllables for loyal 1`] = `
{
  "arpabet": "L OY1 AH0 L",
  "stressLevels": [
    1,
    0,
  ],
  "syllables": [
    "L OY",
    "AH L",
  ],
}
`;

exports[`syllables for ocean 1`] = `
{
  "arpabet": "OW1 SH AH0 N",
  "stressLevels": [
    1,
    0,
  ],
  "syllables": [
    "OW",
    "SH AH N",
  ],
}
`;

exports[`syllables for potato 1`] = `
{
  "arpabet": "P AH0 T EY1 T OW2",
  "stressLevels": [
    0,
    1,
    2,
  ],
  "syllables": [
    "P AH",
    "T EY",
    "T OW",
  ],
}
`;

exports[`syllables for squirrel 1`] = `
{
  "arpabet": "S K W ER1 AH0 L",
  "stressLevels": [
    1,
    0,
  ],
  "syllables": [
    "S K W ER",
    "AH L",
  ],
}
`;

exports[`syllables for suffering 1`] = `
{
  "arpabet": "S AH1 F ER0 IH0 NG",
  "stressLevels": [
    1,
    0,
    0,
  ],
  "syllables": [
    "S AH",
    "F ER",
    "IH NG",
  ],
}
`;

exports[`syllables for throat 1`] = `
{
  "arpabet": "TH R OW1 T",
  "stressLevels": [
    1,
  ],
  "syllables": [
    "TH R OW T",
  ],
}
`;
